packages/features/ee/billing/credit-service.ts

```diff
@@ -0,0 +1,363 @@
+
+      // check if team is still out of credits
+      const teamCredits = await this.getAllCreditsForTeam(teamId);
+      const availableCredits = teamCredits.totalRemainingMonthlyCredits + teamCredits.additionalCredits;
+
+      if (availableCredits > 0) {
+        await CreditsRepository.updateCreditBalance({
+          teamId,
+          data: {
+            limitReachedAt: null,
+            warningSentAt: null,
+          },
+        });
+        return true;
+      }
+    }
+
+    if (userId) {
+      const team = await this.getTeamWithAvailableCredits(userId);
+      return team.availableCredits > 0;
+    }
+
+    return false;
+  }
+
+  async getTeamWithAvailableCredits(userId: number) {
+    const memberships = await MembershipRepository.findAllAcceptedMemberships(userId);
+
+    //check if user is member of team that has available credits
+    for (const membership of memberships) {
+      const creditBalance = await CreditsRepository.findCreditBalance({ teamId: membership.teamId });
+
+      const allCredits = await this.getAllCreditsForTeam(membership.teamId);
+      const limitReached =
+        creditBalance?.limitReachedAt &&
+        dayjs(creditBalance.limitReachedAt).isAfter(dayjs().startOf("month"));
+
+      const availableCredits = allCredits.totalRemainingMonthlyCredits + allCredits.additionalCredits;
+
+      if (!limitReached || availableCredits > 0) {
+        if (limitReached) {
+          await CreditsRepository.updateCreditBalance({
+            teamId: membership.teamId,
+            data: {
+              limitReachedAt: null,
+              warningSentAt: null,
+            },
+          });
+        }
+        return {
+          teamId: membership.teamId,
+          availableCredits,
+          creditType:
+            allCredits.totalRemainingMonthlyCredits > 0 ? CreditType.MONTHLY : CreditType.ADDITIONAL,
+        };
+      }
+    }
+
+    return {
+      teamId: memberships[0].teamId,
+      availableCredits: 0,
+      creditType: CreditType.ADDITIONAL,
+    };
+  }
+
+  /*
+    always returns a team, even if all teams are out of credits
+  */
+  async getTeamToCharge({
+    credits,
+    userId,
+    teamId,
+  }: {
+    credits: number;
+    userId?: number | null;
+    teamId?: number | null;
+  }) {
+    if (teamId) {
+      const teamCredits = await this.getAllCreditsForTeam(teamId);
+      const remaningMonthlyCredits =
+        teamCredits.totalRemainingMonthlyCredits > 0 ? teamCredits.totalRemainingMonthlyCredits : 0;
+      return {
+        teamId,
+        remainingCredits: remaningMonthlyCredits + teamCredits.additionalCredits - credits,
+        creditType: remaningMonthlyCredits > 0 ? CreditType.MONTHLY : CreditType.ADDITIONAL,
+      };
+    }
+
+    if (userId) {
+      const team = await this.getTeamWithAvailableCredits(userId);
+      return { ...team, remainingCredits: team.availableCredits - credits };
+    }
+    return null;
+  }
+
+  private async createExpenseLog(props: {
```

Comment:
I see some operations happening within this function. Do we need to group them in a transaction? I wonder what would happen if one succeeds but another one fails.